### PR TITLE
AP-2497 Alternative alerting mechanism

### PR DESCRIPTION
**Pull request by stephenrichards**
## Alternative alerting mechanism (to Sentry)

[Link to story](https://dsdmoj.atlassian.net/browse/AP-2497)

This PR is a first stab at providing an alternative to Sentry for alerting about exceptions.

- New column on `settings` table to indicate whether or not to use Sentry for alerts
- New class `AlertManager` whose job it is to decide where to send the alert (Sentry or Slack depending on the value in `settings`), and to filter out alerts depending on the type of alert/message or the environment
- New class `SlackAlerter` to send alerts to Slack.  This class responds to the same methods as Sentry, i.e. `.capture exception` and `.capture_message`. Alerts messages and/or exceptions are sent to Slack using email for ease and speed of implementation.
- New mailer `ExceptionAlertMailer` to format the message and send to slack.  

Notes:  
- Slack alert emails are sent immediately, and not via a background job
- All filtering of messages should be done in `AlertManager` where a mechanism exists for this
- Currently, there is no way of capturing the URL of the request that caused the exception - I think this would require the URL to be passed to the `AlertManager#capture_expection` method as a parameter.
- I haven't added the new `settings` field to the Admin page - I expect it to be changed rarely, so we can do it manually in the database.

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/lib/slack_alerter.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class SlackAlerter
+  SLACK_CHANNEL_EMAIL = 'apply-alerts-prod-aaaabbocbcpykkqticw7skurte@mojdt.slack.com'.freeze
```
Comment (colinbruce) [inline app/lib/slack_alerter.rb:2]: Should this be public or regenerated and set in the deployment settings on a per-environment basis?
Wouldn't this mean that a single channel gets _all_ alerts?!

**Review by colinbruce — CHANGES_REQUESTED** [review]
Some issues...

File: config/routes.rb [hunk 2]
```diff
@@ -240,6 +240,12 @@
     end
   end
 
+  # dummy route to test errors in order to see how they are reported in Sentry/Slack
+  # namespace :test do
+  #   resource :generate_error, only: :show
+  # end
```
Comment (mpw5) [inline config/routes.rb:246]: Can these comments be removed?

File: spec/factories/addresses.rb [hunk 3]
```diff
@@ -10,5 +10,9 @@
     trait :is_lookup_used do
       lookup_used { true }
     end
+
```
Comment (mpw5) [inline spec/factories/addresses.rb:13]: This doesn't seem related to the rest of the PR?

File: spec/factories/applicants.rb [hunk 4]
```diff
@@ -10,6 +10,10 @@
       addresses { build_list :address, 1 }
     end
 
+    trait :with_address_for_xml_fixture do
```
Comment (mpw5) [inline spec/factories/applicants.rb:13]: This doesn't seem related to the rest of the PR?

File: spec/services/ccms/requestors/case_add_requestor_spec.rb [hunk 5]
```diff
@@ -23,7 +23,7 @@ module Requestors
         end
         let(:applicant) do
           create :applicant,
-                 :with_address,
+                 :with_address_for_xml_fixture,
```
Comment (mpw5) [inline spec/services/ccms/requestors/case_add_requestor_spec.rb:26]: This doesn't seem related to the rest of the PR?

File: app/lib/alert_manager.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+      alert_klass.capture_message(message)
+    end
+
+    def alert_klass
+      Setting.setting.alert_via_sentry? ? Sentry : SlackAlerter
+    end
+
+    def sendable?(message_or_exception)
+      return false unless HostEnv.environment.in?(%i[production uat])
```
Comment (mpw5) [inline app/lib/alert_manager.rb:20]: Should this also include staging?

**Review by mpw5 — COMMENTED** [review]
Looks good. I've left a few questions and comments. Happy to approve once they've been resolved 👍

File: spec/factories/addresses.rb [hunk 7]
```diff
@@ -10,5 +10,9 @@
...
+    trait :with_address_for_xml_fixture do
+      postcode { 'GH08NY' }
+    end
```
Comment (mpw5) [inline spec/factories/addresses.rb:16]: This doesn't seem related to the rest of the PR?

File: spec/factories/applicants.rb (hunk 4 above)
Comment (stephenrichards) [reply spec/factories/applicants.rb:13]: it is actually - it resolves a flickering test which other changes here made worse - I couldn't work out why. But this ensures that the applicant is created with the correct postcode for the xml fixture file.

File: spec/factories/addresses.rb (hunk 7 above)
Comment (stephenrichards) [reply spec/factories/addresses.rb:16]: it is actually - it resolves a flickering test which other changes here made worse - I couldn't work out why. But this ensures that the applicant is created with the correct postcode for the xml fixture file.

File: config/routes.rb (hunk 2 above)
Comment (stephenrichards) [reply config/routes.rb:246]: removed

File: app/lib/alert_manager.rb (hunk 6 above)
Comment (stephenrichards) [reply app/lib/alert_manager.rb:20]: yes

**Review by mpw5 — APPROVED** [review]
LGTM
